crates/eth_compatibility/src/lib.rs

```diff
@@ -0,0 +1,104 @@
+// Copyright 2018-2021 Parity Technologies (UK) Ltd.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#![cfg_attr(not(feature = "std"), no_std)]
+use ink_env::{
+    DefaultEnvironment,
+    Environment,
+};
+
+/// The ECDSA compressed public key.
+#[derive(Debug, Copy, Clone)]
+pub struct ECDSAPublicKey(pub [u8; 33]);
+
+impl Default for ECDSAPublicKey {
+    fn default() -> Self {
+        Self { 0: [0; 33] }
+    }
+}
+
+impl core::ops::Deref for ECDSAPublicKey {
+    type Target = [u8; 33];
+
+    #[inline]
+    fn deref(&self) -> &Self::Target {
+        &self.0
+    }
+}
+
+impl core::ops::DerefMut for ECDSAPublicKey {
+    #[inline]
+    fn deref_mut(&mut self) -> &mut Self::Target {
+        &mut self.0
+    }
+}
+
+/// The address of an Ethereum account.
+#[derive(Debug, Copy, Clone)]
+pub struct EthereumAddress(pub [u8; 20]);
+
+impl Default for EthereumAddress {
+    fn default() -> Self {
+        Self { 0: [0; 20] }
+    }
+}
+
+impl core::ops::Deref for EthereumAddress {
+    type Target = [u8; 20];
+
+    #[inline]
+    fn deref(&self) -> &Self::Target {
+        &self.0
+    }
+}
+
+impl core::ops::DerefMut for EthereumAddress {
+    #[inline]
+    fn deref_mut(&mut self) -> &mut Self::Target {
+        &mut self.0
+    }
+}
+
+impl ECDSAPublicKey {
+    pub fn to_eth_address(&self) -> EthereumAddress {
+        use ink_env::hash;
+        use secp256k1::PublicKey;
+
+        // Transform compressed public key into uncompressed.
+        let pub_key = PublicKey::parse_compressed(&self.0)
+            .expect("Unable to parse the compressed ECDSA public key");
+        let uncompressed = pub_key.serialize();
+
+        // Hash the uncompressed public key by Keccak256 algorithm.
+        let mut hash = <hash::Keccak256 as hash::HashOutput>::Type::default();
+        // The first byte indicates that the public key is uncompressed.
+        // Let's skip it for hashing the public key directly.
+        ink_env::hash_bytes::<hash::Keccak256>(&uncompressed[1..], &mut hash);
+
+        // Take the last 20 bytes as an Address
+        let mut result = EthereumAddress::default();
+        result.copy_from_slice(&hash[12..]);
+
+        result
+    }
+
+    pub fn to_account_id(&self) -> <DefaultEnvironment as Environment>::AccountId {
```

Comment:
It would be better if we did not only support the `DefaultEnvironment` if that's possible.

Reply:
It can be only `DefaultEnvironment` because it hashes the ECDSA public key and the output is 32 bytes. I copied this logic from the substrate(it is how substrate generates `AccountId` from the ECDSA). 

So the user can't use a custom type of `AccountId` here.

Reply:
If that's the case I'd rather remove `to_account_id` and instead provide a way to extract the `[u8; 32]` so that users can themselves create an `AccountId` from them if necessary.

Reply:
The user already has the ability to get an inner array(it is public).
The idea of this method is to simplify work with default `AccountId`. Because without this implementation the user must check the source code of substrate and find the way how to convert ECDSA public key into `AccountId`.

Reply:
First of all this API is missing proper documentation.
Second I find it very confusing for users that this only works for the default environment.
It would be an improvement if it worked for any environment that uses the default environment account ID instead but that would still lead to confusion.
I'd personally prefer an API that returns something that can easily turned into the default account ID instead, e.g. `[u8; 32]` byte array.